sys/net/application_layer/cord/lc/cord_lc.c

```diff
@@ -0,0 +1,432 @@
+            if (current_attr->value == NULL) {
+                /* skip attributes that have no value */
+                continue;
+            }
+
+            if (!strncmp(current_attr->value, "core.rd-lookup-res",
+                         current_attr->value_len)) {
+                rd->res_lookif = lookif.target;
+                rd->res_lookif_len = lookif.target_len;
+            } else if (!strncmp(current_attr->value, "core.rd-lookup-ep",
+                                current_attr->value_len)) {
+                rd->ep_lookif = lookif.target;
+                rd->ep_lookif_len = lookif.target_len;
+            }
+        }
+    }
+
+    if (!rd->res_lookif && !rd->ep_lookif) {
+        DEBUG("cord_lc: no lookup interfaces found\n");
+        retval = CORD_LC_NORSC;
+    } else {
+        retval = parsed_len;
+    }
+end:
+    mutex_unlock(&_mutex);
+    return retval;
+}
+
+ssize_t cord_lc_raw(const cord_lc_rd_t *rd, unsigned content_format,
+                    unsigned lookup_type, cord_lc_filter_t *filters,
+                    void *result, size_t maxlen)
+{
+    _lock();
+    ssize_t retval = _lookup_raw(rd, content_format, lookup_type, filters,
+                                 result, maxlen);
+    mutex_unlock(&_mutex);
+    return retval;
+}
+
+ssize_t _lookup_result(cord_lc_rd_t *rd, cord_lc_res_t *result,
+                       cord_lc_filter_t *filters, void *buf, size_t maxlen,
+                       unsigned type)
+{
+    int retval;
+
+    unsigned *page_ptr = (type == CORD_LC_EP) ? &rd->ep_last_page : &rd->res_last_page;
+
+    /* encode page number as string */
+    ssize_t len = snprintf(NULL, 0, "%u", *page_ptr);
+    char page_str[len + 1]; /* +1 for '\0' */
+    snprintf(page_str, sizeof(page_str), "%u", *page_ptr);
+    (*page_ptr)++;
+
+    /* Append given filters to default filters (page, count).
+     * If same filter are also specified by filters, assume the RD server will
+     * use the value from last filter in the filter list */
+    clif_attr_t default_attrs[] = DEFAULT_FILTERS(page_str);
+    cord_lc_filter_t *all_filters = &(cord_lc_filter_t) {
+        .array = default_attrs,
+        .len = ARRAY_SIZE(default_attrs),
+        .next = filters,
+    };
+
+    retval = _lookup_raw(rd, COAP_FORMAT_LINK, type, all_filters, buf, maxlen);
+    if (retval < 0) {
+        if (retval == CORD_LC_NORSC) {
+            *page_ptr = 0;
+        }
+        DEBUG("cord_lc: error ep lookup failed\n");
+        return retval;
+    }
+
+    /* parse the result */
+    retval = clif_decode_link(&result->link, result->attrs, result->max_attrs,
+                              buf, retval);
+    if (retval < 0) {
+        DEBUG("cord_lc: no endpoint link found\n");
+        retval = CORD_LC_ERR;
+    }
+    return retval;
+}
+
+ssize_t cord_lc_res(cord_lc_rd_t *rd, cord_lc_res_t *resource,
+                    cord_lc_filter_t *filters, void *buf, size_t maxlen)
+{
+    _lock();
```

Comment:
You can simplify this code a bit: move the `_lock()` and `mutex_unlock()` calls to `_lookup_result()`, and move `cord_lc_res()` and `cord_lc_ep()` into the header file and make them `static inline` as they are simple one-liners then.

Reply:
If I understand you correctly, doing that will require `_lookup_result()` to be moved to the header file too, which is not good I guess because it should only be a helper function and not supposed to be exposed outside of the file.

Reply:
Yes. I don't see any problem with exposing `_lookup_result()` in the public API. There is nothing in this function signature that would prevent users to use it directly, in the contrary, there might even be possible use cases where it could come in handy to programmatically define the type parameter...

But even if you don't like this and leave it like it is, you should move the mutex related lines into `_lookup_result()` to reduce some code duplication...